akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/ActorThreadSpec.scala

```diff
@@ -0,0 +1,298 @@
+import org.scalatest.wordspec.AnyWordSpecLike
+
+object ActorThreadSpec {
+  object Echo {
+    final case class Msg(i: Int, replyTo: ActorRef[Int])
+
+    def apply(): Behavior[Msg] =
+      Behaviors.receiveMessage {
+        case Msg(i, replyTo) =>
+          replyTo ! i
+          Behaviors.same
+      }
+  }
+
+}
+
+class ActorThreadSpec extends ScalaTestWithActorTestKit with AnyWordSpecLike with LogCapturing {
+
+  "Actor thread-safety checks" must {
+
+    "detect illegal access to ActorContext from outside" in {
+      @volatile var context: ActorContext[String] = null
+      val probe = createTestProbe[String]()
+
+      spawn(Behaviors.setup[String] { ctx =>
+        // here it's ok
+        ctx.children
+        context = ctx
+        probe.ref ! "initialized"
+        Behaviors.empty
+      })
+
+      probe.expectMessage("initialized")
+      intercept[UnsupportedOperationException] {
+        context.children
+      }.getMessage should include("[children]")
+
+    }
+
+    "detect illegal access to ActorContext from other thread when processing message" in {
+      val probe = createTestProbe[UnsupportedOperationException]()
+
+      val ref = spawn(Behaviors.receive[CountDownLatch] {
+        case (context, latch) =>
+          Future {
+            try {
+              context.children
+            } catch {
+              case e: UnsupportedOperationException =>
+                probe.ref ! e
+            }
+          }(context.executionContext)
+          latch.await(5, TimeUnit.SECONDS)
+          Behaviors.same
+      })
+
+      val l = new CountDownLatch(1)
+      try {
+        ref ! l
+        probe.receiveMessage().getMessage should include("[children]")
+      } finally {
+        l.countDown()
+      }
+    }
+
+    "detect illegal access to ActorContext from other thread after processing message" in {
+      val probe = createTestProbe[UnsupportedOperationException]()
+
+      val ref = spawn(Behaviors.receive[CountDownLatch] {
+        case (context, latch) =>
+          Future {
+            try {
+              latch.await(5, TimeUnit.SECONDS)
+              context.children
+            } catch {
+              case e: UnsupportedOperationException =>
+                probe.ref ! e
+            }
+          }(context.executionContext)
+
+          Behaviors.stopped
+      })
+
+      val l = new CountDownLatch(1)
+      try {
+        ref ! l
+        probe.expectTerminated(ref)
+      } finally {
+        l.countDown()
+      }
+      probe.receiveMessage().getMessage should include("[children]")
+    }
+
+    "detect illegal access from child" in {
+      val probe = createTestProbe[UnsupportedOperationException]()
+
+      val ref = spawn(Behaviors.receive[String] {
+        case (context, _) =>
+          // really bad idea to define a child actor like this
+          context.spawnAnonymous(Behaviors.setup[String] { _ =>
+            try {
+              context.children
```

Comment:
i guess there's no race in this test as either the setup will happen in parallel and they'd be different threads or after and the parent's current thread would have been cleared

Reply:
That's right. The `checkCurrentActorThread` will detect both cases as either wrong thread or no thread.